### DISCUSSION TITLE
Allow Trailing Comma in Type Implementations

### DISCUSSION THREAD

**Discussion by TheBrambleShark** [dotnet/csharplang#8190] [Language Ideas] [answered]
When working with multiple TFMs, it's possible that you may want to implement interfaces that are only available from certain TFMs. Here's an extreme example from a library I'm developing that targets net461, netstandard2.1, net6, net7, and net8:

```cs
/// <summary>
/// Represents a 9-digit United States social security number.
/// </summary>
[PublicAPI]
[DebuggerDisplay($"{{{nameof(GetDebuggerDisplay)}(),nq}}")]
public readonly struct SocialSecurityNumber
#if NET461_OR_GREATER || NETSTANDARD2_1_OR_GREATER || NET6_0
    : IFormattable,
    IEquatable<SocialSecurityNumber>
#elif NET7
    : IFormattable,
    IEquatable<SocialSecurityNumber>,
    IEqualityOperators<SocialSecurityNumber, SocialSecurityNumber, bool>
#else // NET8_0_OR_GREATER
    : IFormattable,
    IEquatable<SocialSecurityNumber>,
    IEqualityOperators<SocialSecurityNumber, SocialSecurityNumber, bool>,
    IUtf8SpanFormattable,
    IUtf8SpanParsable<SocialSecurityNumber>
#endif
{
    // Specification omitted
}
```

It would be nice to represent this in an abridged format, such as:
```cs
/// <summary>
/// Represents a 9-digit United States social security number.
/// </summary>
[PublicAPI]
[DebuggerDisplay($"{{{nameof(GetDebuggerDisplay)}(),nq}}")]
public readonly struct SocialSecurityNumber
    : IFormattable,
    IEquatable<SocialSecurityNumber>,
#if NET7_0_OR_GREATER
    IEqualityOperators<SocialSecurityNumber, SocialSecurityNumber, bool>,
#endif
#if NET8_0_OR_GREATER
    IUtf8SpanFormattable,
    IUtf8SpanParsable<SocialSecurityNumber>
#endif
{
    // Specification omitted
}
```

This however is currently not possible because C# does not allow trailing commas in the interface implementation list (is there a better name for this?). This leads to needing to re-implement as shown above or do SQL-like commas where the comma lives before the item, assuming style rules don't enforce SA1001.

(While I admit this does look nice, providing the option for the trailing comma allows users to keep compliance with coding styles).
```cs
/// <summary>
/// Represents a 9-digit United States social security number.
/// </summary>
[PublicAPI]
[DebuggerDisplay($"{{{nameof(GetDebuggerDisplay)}(),nq}}")]
public readonly struct SocialSecurityNumber
#pragma warning disable SA1001 // Commas should not be preceeded by whitespace.
    : IFormattable
    , IEquatable<SocialSecurityNumber>
#if NET7_0_OR_GREATER
    , IEqualityOperators<SocialSecurityNumber, SocialSecurityNumber, bool>
#endif
#if NET8_0_OR_GREATER
    , IUtf8SpanFormattable
    , IUtf8SpanParsable<SocialSecurityNumber>
#endif
```

**Comment by siegfriedpammer** [comment]
In that specific case you could put `IFormattable` or a different unconditionally implemented interface last.

Allowing trailing commas everywhere would be nice, but it might cause parsing ambiguities. Also, personally, I'd want the language design team to work on something else.

**Reply by TheBrambleShark** [reply]
Are there any cases you could think of where `public class MyType : IT1, IT2, IT3, { }` could lead to issues?
I'd suspect the compiler after a comma would look for a type name or an instance of `{`, and possibly even remove the last comma for backwards compatibility.

I definitely don't disagree with this being a lower priority change given more impactful features such as DUs up and coming.

**Comment by jnm2** [comment]
It would be interesting if `ValueType` was allowed to be explicitly specified just as in `class Foo : object`.

**Comment by CyrusNajmabadi** [comment, marked as answer]
We have a general issue tracking trailing commas: https://github.com/dotnet/csharplang/issues/391

This should be rolled into that.

**Comment by MillKaDe** [comment]
.. and until trailing commas are allowed, you can avoid the problem (and improve readability) by using `partial` types like this:

Link [SharpLab](https://sharplab.io/#v2:CYLg1APgxMCmBmBLAdrABAOQKIBUBsA+gAwCwAUDAiutjgOzmVKqa4AcxBA8gEoEDiPLAEEcWHuXIABAEwBGSRUTxWOACx453PoJFieaCBFUBlHMIwARYT0syCW3gKGjxh47UKkyUgMxoAJ1gAQ2AAe2QAGwBPNAAHYICAF0RgyLQAZySAgFcAYyS0ExMMNABvcjRAkPCo2JRCjABuSrQ/IpK0AAoGtGQASjQAXgA+TGG+lrIAX3IE5NT0rNyCjtKQNABJLABHHOCk4IAjSNgAHmKMMYqyKqOwsPTtvYPj04uSkYA6Xf3IjO6l0yGQGwzGpSGQ2ByC+zVa7TCADdYAEAog4Gh7o80L80gCumEjgArWAFAD8aEJRMGo0pxLQiABQIyILQADI2TiXv9uiyBlMqu0pHI8JiHukwnEUQcwgFhlCukDggAaNaYmljYKw+WY2ECtr+YWirESqUBGVygCECqVqqBRw1aC1pWturhtwNlORqPR6F6/FgSQAEsEMgALADCYQxXUdGC+AeDocj0fQsamszI8xSaUy2XyhSBG02ADFZQBbA6HE7oG6CuRELZlgKVpLV05fHBhEzZFAAc26wqIFPg5aSqtLFYOAAUAkjfQEKco4nHO93e8gB11R+OGfAVxnGLBkMBlIooMpVAws4kc0t86si205LCcuWUYg8hkvs8/ogktEXBmhaGQfBgdolKqJrXK0AD0sGekaYrYpK0pJLKOqKp0Kpqg6YJOtqkJuvq8GISKyGmmhGGulhpQ4fajrOmgrpHHq5CZlAx6nvA56XrQHBENozh6OIcy3oseYrIWnTFgAqkk8BsCYCTIM2rbtrWrQmls8mKcpwSqVObZvLAnYBNEakHICKlnEc0RJLAYwZHEyCqmEOSFL0ADuqo8DUXB1PpyBnHkYaJGMO4TpZSSzvOcCLnuK6tFUtLxjg5nRbyLluR5aA+WgkWJf0h43gsubLAWapyQpSkqdOiQZCZYEwR6SFApsum1QZ9UBI1NbNV8PUZGmfmhAFMRBbZ9mOZk2VNkZsWIguS77o6qBeWqPTIEkg0NWmzmuUVxXwuR2kdTVQVDU1lzfOl0RDSN/mBTZdkOU5c2Ti2M5zkt8UrXEOXSaUfKDHWVRoIiiSBBMDRmfde1ZYdy6AxDUP8sl0ITOtm1yhSyBoBscDwMEOSREkx0eoKdCBPqmYcVxZ5kEAA===)

```c#
#define NET6_0
#define NET7
#define NET8_0_OR_GREATER

using System;

#if NET461_OR_GREATER || NETSTANDARD2_1_OR_GREATER || NET6_0
public readonly partial struct SSN {
  readonly int N;
  public SSN (int n) => N = n;
}
partial struct SSN : IEquatable<SSN> {
  bool IEquatable<SSN>.Equals (SSN ssn) => N == ssn.N;
  public override bool Equals (object? obj) => obj is SSN ssn && Equals (ssn);
  public static bool operator == (SSN a, SSN b) => a.N == b.N;
  public static bool operator != (SSN a, SSN b) => a.N != b.N;
  public override int GetHashCode () => N.GetHashCode ();
}
partial struct SSN : IFormattable {
  string IFormattable.ToString (string? fmt, IFormatProvider? ifp) => N.ToString (fmt, ifp);
}
#endif

#if NET7
partial struct SSN : System.Numerics.IEqualityOperators<SSN, SSN, bool> {
  // public static bool operator == (SSN a, SSN b) => a.N == b.N;
  // public static bool operator != (SSN a, SSN b) => a.N != b.N;
}
#endif

#if NET8_0_OR_GREATER
partial struct SSN : IUtf8SpanFormattable {
  bool IUtf8SpanFormattable.TryFormat (Span<byte> spn, out int w, ReadOnlySpan<char> fmt, IFormatProvider? ifp)
    => N.TryFormat (spn, out w, fmt, ifp);
}
partial struct SSN : IUtf8SpanParsable<SSN> {
  static SSN IUtf8SpanParsable<SSN>.Parse (ReadOnlySpan<byte> spn, IFormatProvider? ifp) => new SSN (int.Parse (spn, ifp));
  static bool IUtf8SpanParsable<SSN>.TryParse (ReadOnlySpan<byte> spn, IFormatProvider? ifp, out SSN ssn) {
    var r = int.TryParse (spn, ifp, out var n);
    ssn = new SSN (r ? n : default);
    return r;
  }
}
#endif
```
